brave/src/main/java/brave/internal/codec/EntrySplitter.java

```diff
@@ -0,0 +1,315 @@
+     * <pre>{@code
+     * String key = input.substring(beginKey, endKey);
+     * String value = input.substring(beginValue, endValue);
+     * }</pre>
+     *
+     * @param target receiver of parsed entries
+     * @param input string including data to parse
+     * @param beginKey begin index of the entry's key in {@code input}, inclusive
+     * @param endKey end index of the entry's key in {@code input}, exclusive
+     * @param beginValue begin index of the entry's value in {@code input}, inclusive
+     * @param endValue end index of the entry's value in {@code input}, exclusive
+     * @return true if we reached the {@code endIndex} without failures.
+     */
+    boolean onEntry(
+        T target, String input, int beginKey, int endKey, int beginValue, int endValue);
+  }
+
+  final char keyValueSeparator, entrySeparator;
+  int maxEntries;
+  final boolean trimOWSAroundEntrySeparator, trimOWSAroundKeyValueSeparator;
+  final boolean keyValueSeparatorRequired, shouldThrow;
+  final String missingKey, missingKeyValueSeparator, overMaxEntries;
+
+  EntrySplitter(Builder builder) {
+    keyValueSeparator = builder.keyValueSeparator;
+    entrySeparator = builder.entrySeparator;
+    maxEntries = builder.maxEntries;
+    trimOWSAroundEntrySeparator = builder.trimOWSAroundEntrySeparator;
+    trimOWSAroundKeyValueSeparator = builder.trimOWSAroundKeyValueSeparator;
+    keyValueSeparatorRequired = builder.keyValueSeparatorRequired;
+    shouldThrow = builder.shouldThrow;
+    missingKey = "Invalid input: no key before '" + keyValueSeparator + "'";
+    missingKeyValueSeparator =
+        "Invalid input: missing key value separator '" + keyValueSeparator + "'";
+    overMaxEntries = "Invalid input: over " + maxEntries + " entries";
+  }
+
+  /**
+   * @param handler parses entries emitted upon success
+   * @param target receiver of parsed entries
+   * @param input string including data to parse
+   * @return true if we reached the {@code endIndex} without failures.
+   */
+  public <T> boolean parse(Handler<T> handler, T target, String input) {
+    if (input == null) throw new NullPointerException("input == null");
+    return parse(handler, target, input, 0, input.length());
+  }
+
+  /**
+   * @param handler parses entries emitted upon success
+   * @param target receiver of parsed entries
+   * @param input string including data to parse
+   * @param beginIndex begin index of the {@code input}, inclusive
+   * @param endIndex end index of the {@code input}, exclusive
+   * @return true if we reached the {@code endIndex} without failures.
+   */
+  public <T> boolean parse(
+      Handler<T> handler, T target, String input, int beginIndex, int endIndex) {
+    if (handler == null) throw new NullPointerException("handler == null");
+    if (target == null) throw new NullPointerException("target == null");
+    if (input == null) throw new NullPointerException("input == null");
```

Comment:
Should we also use it as `IllegalArgumentException` in those argument validations here? When I pass an invalid argument to a method I expect the method raises some validation error about those instead off give feedback about the invalid arguments wrapped in an NPE error.

Reply:
On this one I don't agree. the only reason we explicitly raise them here is to prevent confusion later when java would do it for you, but without a message.

If it were different, ex two Nullable params and one must be non-null, yeah that would be an IAE.

Make sense?

Reply:
it makes senses but at the same time, it looks like we are translating the NPE to a more readable NPE. But in the end, the method doesn't work with one of those arguments is not null. So there is a dependency on this method works that is why I think the IAE makes more sense. Imagining to add `@NotNull` to the arguments so we would see a sort of `Argument for @NotNull parameter 'b'`.

Reply:
I found it a bit weird a long time ago but think it's general convention to use NPE for these, codified into Java 8 with `requireNonNull` which throws NPE.

Reply:
argg.... there is not consistency between `dosomething(@NotNull String a)` x `requireNonNull(a)` oO
So, it's fine either way. no argument for now anymore. :)